src/Lean/Elab/Deriving/Util.lean

```diff
@@ -36,59 +41,340 @@ open TSyntax.Compat in
 /-- Return implicit binder syntaxes for the given `argNames`. The output matches `implicitBinder*`.
 
 For example, ``#[`foo,`bar]`` gives `` `({foo} {bar})``. -/
-def mkImplicitBinders (argNames : Array Name) : TermElabM (Array (TSyntax ``Parser.Term.implicitBinder)) :=
+def mkImplicitBinders (argNames : Array Name) : TermElabM (Array (TSyntax ``Parser.Term.implicitBinder)) := do
   argNames.mapM fun argName =>
     `(implicitBinderF| { $(mkIdent argName) })
 
+/--
+    Represents `ind params1 ... paramsn`, where `paramsi` is either a nested occurence, a constant name, or a free variable
+
+    Example :
+    ```
+    inductive Foo (α β γ: Type) : Type → Type
+
+    inductive Bar
+        | foo : A -> B -> Foo A (Option Bar) B Nat → Bar
+
+    ```
+    this nested occurence `Foo A (Option Bar) B Nat` is encoded as
+    ```lean
+      node Foo #[
+        .inr (.bvar 2),
+        .inl (node Option #[.inr (.leaf Bar #[])]),
+        .inr (.bvar 1),
+        ]
+    ```
+
+    Remark 1 : Since an inductive can only be nested as a parameter, not an index, of another inductive type,
+    we assume `params.size == ind.numParams`
+
+    Remark 2 : Free variables are abstracted away as bound variables in nested occurences.
+    This is useful when trying to delete duplicate occurences, since the check is now purely syntactical.
+  -/
+inductive NestedOccurence : Type :=
+  | node (ind : InductiveVal) (params : Array (NestedOccurence ⊕ Expr))
+  | leaf (ind : InductiveVal) (fvars : Array Expr)
```

Comment:
This seems to the be core concept in this PR. I wonder if we can improve the naming a bit – if `NestedOccurence` shows up in the type signature, it doesn’t really say much.

Also I am not sure if I would call `Foo A (Option Bar) B Nat` a nested occurrence (of `Bar`) – the nested occurrence might be the the `Bar` inside it.

But I see that it’s not easy to come up with a good name. The concept behind it is “in the delaboration of nested inductiction, this corresponds to one of the auxillary inductive definition”, right?

In [Lean.Elab.Structural.IndGroupInst.nestedTypeFormers](https://leanprover-community.github.io/mathlib4_docs/Lean/Elab/PreDefinition/Structural/IndGroupInfo.html#Lean.Elab.Structural.IndGroupInst.nestedTypeFormers) I call them “nested type formers”. Not if that helps.

Maybe `IndTypeFormer`?  At least it has `Ind` in the name.

I’m also a bit surprised that you have to introduce a tree-shaped data type for it. Why not keep the `Expr` around, and traverse and take apart as needed?

(also, `occurrence` with two `r`, right?)

Reply:
>  if NestedOccurence shows up in the type signature, it doesn’t really say much. 

Right, I agree that this needs renaming

> I am not sure if I would call `Foo A (Option Bar) B Nat` a nested occurrence

I followed the naming done in [this comment](https://github.com/leanprover/lean4/blob/f989520d2b17ec2fe987a69fee44a5272be1ff10/src/kernel/inductive.cpp#L873C71-L873C88) in the kernel, but I'm not sure I agree with that phrasing either. 

> The concept behind it is “in the delaboration of nested inductiction, this corresponds to one of the auxillary inductive definition”, right?

Yes, the idea behind this PR is to take inspiration from the kernel code which converts nested inductives into mutual inductives, so that we may mutually define the definitions necessary for deriving the instances.

> Maybe `IndTypeFormer` ?

Works for me ! I'll rename it.

> I’m also a bit surprised that you have to introduce a tree-shaped data type for it. Why not keep the Expr around, and traverse and take apart as needed ?

I first started my work on this PR about a year ago, so my memory is a bit fuzzy regarding my thought process back then. I definitely did think that this would allow me to work more easily on this, but it is certainly possible to get rid of this tree structure entirely and work on exprs instead, although that would require a fair bit of rewriting.

Reply:
I wonder if, now that it’s there, you could use the [Lean.Elab.Structural.IndGroupInst](https://leanprover-community.github.io/mathlib4_docs/Lean/Elab/PreDefinition/Structural/IndGroupInfo.html#Lean.Elab.Structural.IndGroupInst) infrastructure instead. This is what’s used by the structural recursion compiler, which is what you are targetting, so maybe that's natural and easier.

Once you instantiate the `IndGroupInst` with your universe parameters and free variables for the inductive parameters, you can get all typeformers in the group. And when you look at the argument of a constructor, you can find out which type former it’s an instance of, and with which index instantiations, like so:

https://github.com/leanprover/lean4/blob/e9e858a4484905a0bfe97c4f05c3924ead02eed8/src/Lean/Elab/PreDefinition/Structural/FindRecArg.lean#L198-L200

Reply:
This might indeed work. One difference is that I can't exactly rely on `IndGroupInst.nestedTypeFormers` to make this work. In particular, I also need to look for nested instances inside the "auxiliary inductive definitions" generated, until I reach a fixpoint. For example when working with `inductive Tree where | node : Array Tree → Tree`, I'll not only need a nested type former for `Array`, but also one for `List`.